File: 03-luhn-kotlin/src/de/oio/luhn/Luhn.kt
package de.oio.luhn

fun main(args: Array<String>) {
	println(isValid(378282246310005)) // true
	println(isValid(76009244561)) // false
}

fun isValid(number: Long): Boolean {
	var even = false
	return number.toString()
			.split("")
			.reversed()
			.filter { !it.isBlank() }
			.map { it.toInt() }
			.map { even = !even; if (even) it else it * 2 }
			.map { if (it > 9) it - 9 else it }
			.sum() % 10 == 0
}

File: 03-luhn-javascript/luhn.js
console.log(isValid(378282246310005)) // true
console.log(isValid(76009244561)) // false

function isValid(number) {
    let even = false
    return number.toString()
        .split('')
        .reverse()
        .map(c => parseInt(c))
        .map(d => (even = !even) ? d : d < 5 ? d * 2 : (d - 5) * 2 + 1)
        .reduce((a, b) => a + b, 0)
        % 10 === 0
}

